src/pk_encryption_circuit.rs

```diff
@@ -0,0 +1,714 @@
+
+        // The parameters are auto configured by halo2 lib to fit all the columns into the `k`-sized table
+        let rlc_circuit_params = rlc_circuit.0.calculate_params(Some(9));
+
+        // 4. Generate the verification key and the proving key
+        let vk = keygen_vk(&kzg_params, &rlc_circuit).unwrap();
+        let pk = keygen_pk(&kzg_params, vk, &rlc_circuit).unwrap();
+        let break_points = rlc_circuit.0.builder.borrow().break_points();
+        drop(rlc_circuit);
+
+        // 5. Generate the proof, here we pass the actual inputs
+        let mut proof_gen_builder: RlcCircuitBuilder<Fr> =
+            RlcCircuitBuilder::from_stage(CircuitBuilderStage::Prover, 0)
+                .use_params(rlc_circuit_params);
+        proof_gen_builder.base.set_lookup_bits(k - 1);
+        proof_gen_builder.base.set_instance_columns(1);
+
+        let file_path = "src/data/pk_enc_data/pk_enc_1024_15x60_65537.json";
+        let mut file = File::open(file_path).unwrap();
+        let mut data = String::new();
+        file.read_to_string(&mut data).unwrap();
+        let pk_enc_circuit = serde_json::from_str::<BfvPkEncryptionCircuit>(&data).unwrap();
+        let instances = pk_enc_circuit.instances();
+
+        let rlc_circuit = RlcExecutor::new(proof_gen_builder, pk_enc_circuit.clone());
+
+        rlc_circuit
+            .0
+            .builder
+            .borrow_mut()
+            .set_break_points(break_points);
+        let proof = gen_proof_with_instances(&kzg_params, &pk, rlc_circuit, &[&instances[0]]);
+
+        // 6. Verify the proof
+        check_proof_with_instances(&kzg_params, pk.get_vk(), &proof, &[&instances[0]], true);
+    }
+
+    #[test]
+    pub fn test_pk_enc_invalid_range() {
+        // 1. Define the inputs of the circuit
+        let file_path = "src/data/pk_enc_data/pk_enc_1024_15x60_65537.json";
+        let mut file = File::open(file_path).unwrap();
+        let mut data = String::new();
+        file.read_to_string(&mut data).unwrap();
+        let mut pk_enc_circuit = serde_json::from_str::<BfvPkEncryptionCircuit>(&data).unwrap();
+        let instances = pk_enc_circuit.instances();
+
+        // 2. Invalidate the circuit by setting the value of a coefficient of the polynomial `r1is[0]` to be out of range
+        let out_of_range_coeff = R1_BOUNDS[0] + 1;
+        pk_enc_circuit.r1is[0][0] = out_of_range_coeff.to_string();
+
+        // 3. Build the circuit for MockProver
+        let rlc_circuit_params = test_params();
+        let mut mock_builder: RlcCircuitBuilder<Fr> =
+            RlcCircuitBuilder::from_stage(CircuitBuilderStage::Mock, 0)
+                .use_params(rlc_circuit_params.clone());
+        mock_builder.base.set_lookup_bits(8); // Set the lookup bits to 8
+
+        mock_builder.base.set_instance_columns(1);
+
+        let rlc_circuit = RlcExecutor::new(mock_builder, pk_enc_circuit);
+
+        // 4. Run the mock prover
+        let invalid_mock_prover = MockProver::run(
```

Comment:
This test is not actually failing for the reason that it is supposed to fail. Instead it is failing because `NOT ENOUGH ADVICE COLUMNS`, you need to increase the `k` factor to 22 and you will see that the error disappear. Also please check the invalid range also on p2i and p1i